### PR TITLE
JS: detect fs modules that pass through a reduce call

### PR DESCRIPTION
**Pull request by erik-krogh**
Detects the sink in CVE-2020-26299

[An evaluation on `TaintedPath.ql`](https://github.com/dsp-testing/erik-krogh-dca/tree/run/moreFS-default-taintedPath/reports) was uneventful.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll [hunk 1]
```diff
@@ -478,6 +478,28 @@ module NodeJSLib {
               DataFlow::moduleImport("util-promisifyall")
             ].getACall()
         )
+        or
+        // const fs = require('fs');
+        // module.exports = methods.reduce((obj, method) => {
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll:483]:
```suggestion
        // let fs_copy = methods.reduce((obj, method) => {
```
We do not use `module.exports = ` in the implementation, yet it is a special enough construct that I expected to find a reference to it in the implementation when it is used in the example.

File: javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll [hunk 2]
```diff
@@ -478,6 +478,28 @@ module NodeJSLib {
...
+        //  obj[method] = fs[method];
+        //  return obj;
+        // }, {});
+        t.continue() = t2 and
+        exists(
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll:488]: This step through `reduce` seems general enough that it belongs in its own reusable predicate, could the  `fsModule` and `maybePromisified` predicates be factored out?

**Review by esbena — COMMENTED** [review]
LGTM, although the many additional `fs`-module steps are starting to be a bit worrying. 
Should we perhaps try to quantify how many/few other modules that get sent through similar hoops?

File: javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll (hunk 2 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll:488]: We could refactor the second half of `fsModule` into a `maybePromisefyObject`.   
But I think that is something we can look into when we find another model that needs the same treatment.

**Review by esbena — APPROVED** [review]

**Pull request merged**